File: DesignPatterns/2-strategyPattern/golang/context.go
/*
 * File: context.go
 * Created Date: 2021-12-12 04:25:39
 * Author: ysj
 * Description:  策略上下文
 */

package main

type PriceCalculationContext struct {
	strategy IPriceCalculationStrategy
}

func NewPriceCalculationContext(strategy IPriceCalculationStrategy) *PriceCalculationContext {
	return &PriceCalculationContext{
		strategy: strategy,
	}
}

func (p *PriceCalculationContext) calculatePrice(originPrice float64) float64 {
	return p.strategy.calculatePrice(originPrice)
}

File: DesignPatterns/2-strategyPattern/typescript/context.ts
/**
 * -------------------------------------------------------
 * File: context.ts
 * Created Date: 2021-12-12 01:46:28
 * Author: ysj
 * Description: 策略上下文
 * -------------------------------------------------------
 */

import { IPriceCalculationStrategy } from './strategy';

export default class PriceCalculationContext {
  private strategy: IPriceCalculationStrategy;
  constructor(strategy: IPriceCalculationStrategy) {
    this.strategy = strategy;
  }
  calculatePrice(originPrice: number): number {
    const totalPrice = this.strategy.calculatePrice(originPrice);
    return totalPrice;
  }
}
